test/TestRoot.js

```diff
@@ -0,0 +1,110 @@
+const Root = artifacts.require('./Root.sol');
+const DNSSEC = artifacts.require('./mocks/DummyDNSSEC.sol');
+const ENS = artifacts.require('./ENSRegistry.sol');
+
+const utils = require('./helpers/Utils.js');
+const namehash = require('eth-ens-namehash');
+const dns = require('../lib/dns.js');
+
+contract('Root', function(accounts) {
+
+    let node;
+    let ens, dnssec, root;
+
+    let now = Math.round(new Date().getTime() / 1000);
+
+    beforeEach(async function() {
+        node = namehash.hash('eth');
+
+        ens = await ENS.new();
+        dnssec = await DNSSEC.new();
+        root = await Root.new(ens.address, dnssec.address);
+
+        await ens.setSubnodeOwner(0, web3.sha3('eth'), root.address, {from: accounts[0]});
+        await ens.setOwner(0, root.address);
+    });
+
+    describe('setSubnodeOwner', async () => {
+
+        it('should fail when trying to set subnode owner for non root domain', async () => {
+            try {
+                await root.setSubnodeOwner(web3.sha3('eth'), '0x123', accounts[1], {from: accounts[0]});
+            } catch (error) {
+                return utils.ensureException(error);
+            }
+        });
+
+        it('should allow setting subnode when trying to owner for root domain', async () => {
+            await root.setSubnodeOwner(0, web3.sha3('eth'), accounts[1], {from: accounts[0]});
+            assert.equal(accounts[1], await ens.owner(node));
+        });
+    });
+
+    it('should allow transferring ownership of the root node', async () => {
+        assert.equal(root.address, await ens.owner(0));
+        await root.transferRoot(accounts[1]);
+        assert.equal(accounts[1], await ens.owner(0));
+    });
+
+    describe('registerTLD', async () => {
+
+        it('allows registering a TLD on ENS with a custom address', async () => {
+            let proof = dns.hexEncodeTXT({
+                name: '_ens.test.',
+                klass: 1,
+                ttl: 3600,
+                text: ['a=' + accounts[0]]
+            });
+
+            await dnssec.setData(
+                16,
+                dns.hexEncodeName('_ens.test.'),
+                now,
+                now,
+                proof
+            );
+
+            await root.registerTLD(dns.hexEncodeName('test.'), proof);
+
+            assert.equal(await ens.owner(namehash.hash('test')), accounts[0]);
+        });
+
+        it('should set TLD owner to default registrar when 0x0 is provided', async () => {
+
+            let proof = dns.hexEncodeTXT({
```

Comment:
This doesn't seem like a thing we should do. If there's *no* text record, it should set it to the default, but if there's any kind of matching text record, we should set it to that.

Reply:
Even for a 0x0 record?

Reply:
I don't think we should allow for that out of security tbh

Reply:
If a domain wants to set the registrar address to 0, meaning nobody can register anything, I think that's fine.

Reply:
Ok, I'll need to think about how to do it in the new code that shares the library.